src/ds++/test/cxx11example_move_semantics.cc

```diff
@@ -0,0 +1,202 @@
+
+struct B {
+
+  /*!
+   * \brief Constructor
+   * \param[in] v_in v_in is lvalue; its type is rval ref
+   */
+  explicit B(std::vector<double> &&v_in)
+      : v_(std::move(v_in)) /* std::move casts to rval, move ctor called */
+  {
+    /* empty */
+  }
+
+  // member data
+  std::vector<double> v_;
+};
+
+//----------------------------------------------------------------------------//
+/*! \breif Demonstration of move semantics
+ *
+ * 1. Create a vector
+ * 2. Attempt to construct a class (A), demonstrate that ownership is not
+ *    transferred.
+ * 3. Attempt to construct another class (B), demonstrate that ownership is
+ *    transferred.
+ */
+//----------------------------------------------------------------------------//
+void move_semantics_example(rtt_dsxx::UnitTest &ut) {
+  using namespace std;
+
+  cout << ">>> Begin demonstration...\n" << endl;
+
+  // Create a complex object.
+  cout << "Create a vector v1.\n";
+  vector<double> v1 = {1, 2, 3};
+  report_memory_locations(v1, "v1");
+
+  auto const v1_loc = &v1;
+  auto v1_data_loc = &v1[0];
+
+  // Case 1:
+  // Create an object, attempt to transfer ownership from v1 to a.
+  // This will fail.
+  cout << "\nCreate an instantiation of A that owns a copy of v1.";
+  A a(move(v1));
+  cout << "\nAfter call to A::ctor\n";
+  report_memory_locations(v1, "v1");
+  report_memory_locations(a.v_, "a.v_");
+
+  // v1 remains unchanged! (not the behavior we want)
+  if (v1_loc != &v1)
+    ITFAILS;
+  if (v1_data_loc != &v1[0])
+    ITFAILS;
+  else
+    PASSMSG("v1 remains unchanged! (but this is not the behavior we want).");
+  if (v1_loc == &a.v_)
+    ITFAILS;
+  if (v1_data_loc == &a.v_[0])
+    ITFAILS;
+  else
+    PASSMSG("Object 'a' has made a copy of v1 (ownership not transfered).");
+  if (rtt_dsxx::soft_equiv(v1.begin(), v1.end(), a.v_.begin(), a.v_.end()))
+    PASSMSG("a.v_ matches v1.");
+  else
+    FAILMSG("A's constructor did not copy the vector's data correctly.");
+
+  // change the data in the vector. Print the new state.
+  cout << "\nExamine the behavior of 'swap'.\n";
+  a.v_[0] = 4;
+  a.v_[1] = 5;
+  a.v_[2] = 6;
+  a.v_.push_back(7); // may force the vector to resize (&a.v_[0] will change!)
+  a.v_.swap(v1);
+  cout << "After call to a.v_.swap(v1):\n";
+  report_memory_locations(v1, "v1");
+  report_memory_locations(a.v_, "a.v_");
+
+  // a.v_ should be the memory location of v1's original data.
+  if (v1_data_loc != &a.v_[0])
+    ITFAILS;
+
+  // The v1 vector's data store may have changed location when we used
+  // 'push_back' above.
+  v1_data_loc = &v1[0];
+
+  // location of v1 remains unchanged (even though it's underlying data
+  // container may be new).
+  if (v1_loc != &v1)
+    ITFAILS;
+
+  // Case 2:
+  // Create an object, attempt to transfer ownership from v1 to b.
+  // This works.
+  cout << "\nCreate an instantiation of B that takes ownership of v1's data.";
+  B b(std::move(v1));
```

Comment:
Since `using namespace std` has been specified, `std::move` could be shortened to `move` (feel free to ignore this suggestion, of course).